### PR TITLE
Allow wildcard fqdn's for HTTPProxy resources

### LINKED ISSUE #1228: Please add support for wildcard virtual hosts
Hi,

I need multiple virtual host fqdn routed to the same backend service, example:
devxnew-workspaces-ws-8zdgz.istio-wing.dev1.sapwebide.net.sap
devxnew-workspaces-ws-8zdgz-app1.istio-wing.dev1.sapwebide.net.sap
devxnew-workspaces-ws-8zdgz-app2.istio-wing.dev1.sapwebide.net.sap
devxnew-workspaces-ws-8zdgz-app3.istio-wing.dev1.sapwebide.net.sap
devxnew-workspaces-ws-8zdgz-app4.istio-wing.dev1.sapwebide.net.sap
devxnew-workspaces-ws-8zdgz-app5.istio-wing.dev1.webide.net.sap

These virtual host should be replaced with one 
devxnew-workspaces-ws-8zdgz*.istio-wing.dev1.webide.net.sap

**Describe the solution you'd like**
virtual host fqdn support regexp
example: aaa*.dev.one.com

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Contour version:
- Kubernetes version: (use `kubectl version`):
- Kubernetes installer & version:
- Cloud provider or hardware configuration:
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by stevesloka**
Fixes #1228

Signed-off-by: Steve Sloka <slokas@vmware.com>

### CODE REVIEW

File: test/e2e/httpproxy/fqdn_test.go [hunk 1]
```diff
@@ -81,4 +83,99 @@ func testWildcardFQDN(namespace string) {
 		err := f.CreateHTTPProxy(p)
 		require.NotNil(t, err, "Expected invalid wildcard to be rejected.")
 	})
+
+	Specify("wildcard routing works", func() {
+		t := f.T()
+
+		f.Fixtures.Echo.Deploy(namespace, "echo-slash-default")
```
Comment (sunjayBhatia) [inline test/e2e/httpproxy/fqdn_test.go:90]: nit: maybe name these with a little more informative names, just so the test cases below read a little easier?

File: test/e2e/httpproxy/fqdn_test.go [hunk 2]
```diff
@@ -81,4 +83,99 @@ func testWildcardFQDN(namespace string) {
...
+		f.CreateHTTPProxyAndWaitFor(proxyFullFQDN, httpProxyValid)
+		f.CreateHTTPProxyAndWaitFor(proxyFullFQDNSubdomain, httpProxyValid)
+
+		cases := map[string]string{
+			"projectcontour.io":                         "ingress-conformance-echo",
+			"www.projectcontour.io":                     "echo-slash-default",
+			"bar.projectcontour.io":                     "ingress-conformance-bar",
+			"foo.projectcontour.io":                     "echo-slash-default",
+			"foo.bar.projectcontour.io":                 "echo-slash-default",
+			"i.n.g.r.e.s.s.r.u.l.e.s.projectcontour.io": "echo-slash-default",
```
Comment (sunjayBhatia) [inline test/e2e/httpproxy/fqdn_test.go:161]: do we want this to match the wildcard? that would be different behavior than Ingress and Gateway API which only allow a single DNS label to match: https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.Hostname

we used this to in the Ingress and Gateway API processors: https://github.com/projectcontour/contour/blob/0d824ba6d0bf3a4a97ee2024b4a45fea21b278ec/internal/dag/ingress_processor.go#L263-L278

Comment (stevesloka) [reply test/e2e/httpproxy/fqdn_test.go:161]: yeah can match that, no reason to deviate.

Comment (stevesloka) [reply test/e2e/httpproxy/fqdn_test.go:161]: you're ruining my fun @sunjayBhatia! :grin:

**Review by youngnick — APPROVED** [review]
LGTM, nice work.

(once the tests are fixed, obviously).

**Comment by sunjayBhatia** [comment]
https://github.com/projectcontour/contour/pull/4147 might be a fix to the test flakes here

File: test/e2e/httpproxy/fqdn_test.go [hunk 3]
```diff
@@ -49,36 +51,135 @@ func testWildcardSubdomainFQDN(namespace string) {
...
+			"bar.projectcontour.io": {
+				ServiceName:   "barprojectcontourio",
+				ShouldSucceed: true,
+			},
+			"foo.projectcontour.io": {
+				ServiceName:   "wildcardprojectcontourio",
+				ShouldSucceed: true,
+			},
+			"bar.foo.projectcontour.io": {
+				ServiceName:   "wildcardprojectcontourio",
```
Comment (sunjayBhatia) [inline test/e2e/httpproxy/fqdn_test.go:147]: tiny nit: could remove since we don't expect a service here

**Review by sunjayBhatia — APPROVED** [review]

File: internal/dag/httpproxy_processor.go [hunk 4]
```diff
@@ -687,6 +688,24 @@ func (p *HTTPProxyProcessor) computeRoutes(
 		if len(r.Clusters) == 0 {
 			r.DirectResponse = directResponse(http.StatusServiceUnavailable)
 		}
+
+		// If we have a wildcard match, add a header match regex rule to match the
+		// hostname so we can be sure to only match one DNS label. This is required
+		// as Envoy's virtualhost hostname wildcard matching can match multiple
+		// labels. This match ignores a port in the hostname in case it is present.
+		if strings.HasPrefix(rootProxy.Spec.VirtualHost.Fqdn, "*.") {
+			r.HeaderMatchConditions = []HeaderMatchCondition{
```
[the remark this answers is not in the document]
Comment (stevesloka) [reply internal/dag/httpproxy_processor.go]: :+1:

**Comment by stevesloka** [comment]
> #4147 might be a fix to the test flakes here

That still may help, but what I found was I think there was a timing issue since many of our tests use `projectcontour.io`, the wildcard `*.projectcontour.io` was left around and causing routes to direct to the wildcard routes. Changing to a non-projectcontour.io domain seems to have fixed that specific flake.

**Pull request merged**